### PR TITLE
Improved issubset

### PR DESCRIPTION
**Pull request by abhinavmehndiratta**
Makes use of the fact that neighbors are sorted and unique for SimpleGraphs.
Benchmarks - 
```julia
julia> g = loadsnap(:soc_slashdot0902_u)
{82168, 582533} undirected simple Int64 graph

julia> for i = 1:1000
           z = collect(edges(g))
           rem_edge!(g, z[rand(1:length(z))])
       end

julia> g
{82168, 581533} undirected simple Int64 graph

julia> h = loadsnap(:soc_slashdot0902_u)
{82168, 582533} undirected simple Int64 graph

julia> issubset(g, h)
true

julia> @benchmark issubset(g, h)	#new
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     8.628 ms (0.00% GC)
  median time:      8.769 ms (0.00% GC)
  mean time:        8.840 ms (0.00% GC)
  maximum time:     10.618 ms (0.00% GC)
  --------------
  samples:          565
  evals/sample:     1

julia> @benchmark issubset(g, h)	#old
BenchmarkTools.Trial: 
  memory estimate:  52.53 MiB
  allocs estimate:  1164076
  --------------
  minimum time:     272.311 ms (1.31% GC)
  median time:      277.995 ms (1.24% GC)
  mean time:        284.930 ms (3.40% GC)
  maximum time:     367.033 ms (22.04% GC)
  --------------
  samples:          18
  evals/sample:     1

```

### CODE REVIEW

**Comment by sbromberger** [comment]
I think this is fine. What I'd like to see in addition is some explicit documentation, perhaps in the [Developer Notes](http://juliagraphs.github.io/LightGraphs.jl/latest/developing.html) section or a new subsection off of that section that specifies this constraint.

File: src/SimpleGraphs/SimpleGraphs.jl [hunk 1]
```diff
@@ -76,8 +76,25 @@ add_edge!(g::AbstractSimpleGraph, x, y) = add_edge!(g, edgetype(g)(x, y))
...
-    (nv(g) <= nv(h)) && issubset(edges(g), edges(h))
+function issubset(g::T, h::T) where T <: AbstractSimpleGraph
+	(nv(g) <= nv(h)) || return false
+	for u in vertices(g)
+		u_nbrs_g = neighbors(g, u)
+		u_nbrs_h = neighbors(h, u)
+		p = 1
+		len_u_nbrs_g = length(u_nbrs_g)
+		(len_u_nbrs_g == 0) && continue
+		(len_u_nbrs_g > length(u_nbrs_h)) && return false
```
Comment (sbromberger) [inline src/SimpleGraphs/SimpleGraphs.jl]: No parens needed here or on the line above.

File: src/SimpleGraphs/SimpleGraphs.jl [hunk 2]
```diff
@@ -76,8 +76,25 @@ add_edge!(g::AbstractSimpleGraph, x, y) = add_edge!(g, edgetype(g)(x, y))
...
+		for v in u_nbrs_h
+			if v == u_nbrs_g[p]
+				(p == len_u_nbrs_g) && break
+				p += 1
+			end
+		end
+		(p == len_u_nbrs_g) || return false
```
Comment (sbromberger) [inline src/SimpleGraphs/SimpleGraphs.jl]: see above

File: src/SimpleGraphs/SimpleGraphs.jl [hunk 3]
```diff
@@ -76,8 +76,25 @@ add_edge!(g::AbstractSimpleGraph, x, y) = add_edge!(g, edgetype(g)(x, y))
...
-issubset(g::T, h::T) where T<:AbstractSimpleGraph =
-    (nv(g) <= nv(h)) && issubset(edges(g), edges(h))
+function issubset(g::T, h::T) where T <: AbstractSimpleGraph
+	(nv(g) <= nv(h)) || return false
+	for u in vertices(g)
+		u_nbrs_g = neighbors(g, u)
+		u_nbrs_h = neighbors(h, u)
+		p = 1
+		len_u_nbrs_g = length(u_nbrs_g)
+		(len_u_nbrs_g == 0) && continue
```
Comment (sbromberger) [inline src/SimpleGraphs/SimpleGraphs.jl]: move this before the `u_nbrs_h = neighbors` so that we don't allocate if we're just going to continue.

**Comment by abhinavmehndiratta** [comment]
> I think this is fine. What I'd like to see in addition is some explicit documentation, perhaps in the [Developer Notes](http://juliagraphs.github.io/LightGraphs.jl/latest/developing.html) section or a new subsection off of that section that specifies this constraint.

But this is just for `SimpleGraphs`, why do we need to specify it in the Developer section ?

**Comment by sbromberger** [comment]
> But this is just for SimpleGraphs, why do we need to specify it in the Developer section ?

It's not. It's for `AbstractSimpleGraph`, and someone might write a concrete type inheriting from ASG and would need to know that `neighbors` needed to be ordered.

**Comment by sbromberger** [comment]
Also, you might be able to speed up `SimpleGraph` since I think you can stop the check when the first vertex is greater than or equal to the second,.

**Comment by abhinavmehndiratta** [comment]
> It's not. It's for `AbstractSimpleGraph`, and someone might write a concrete type inheriting from ASG and would need to know that `neighbors` needed to be ordered.

Okay, so I will specify it in the docs under Developer notes.

File: src/SimpleGraphs/SimpleGraphs.jl [hunk 4]
```diff
@@ -76,8 +76,25 @@ add_edge!(g::AbstractSimpleGraph, x, y) = add_edge!(g, edgetype(g)(x, y))
...
+        len_u_nbrs_g = length(u_nbrs_g)
+        len_u_nbrs_g == 0 && continue
+        u_nbrs_h = neighbors(h, u)
+        p = 1
+        len_u_nbrs_g > length(u_nbrs_h) && return false
+        for v in u_nbrs_h
+            if v == u_nbrs_g[p]
+            p == len_u_nbrs_g && break
+            p += 1
+            end
```
Comment (simonschoelly) [inline src/SimpleGraphs/SimpleGraphs.jl]: Can you put an `@inbounds` in front of the array access, and correct the code indentation?

**Comment by abhinavmehndiratta** [comment]
The build failure is unrelated, please restart it.

**Comment by matbesancon** [comment]
@abhinavmehndiratta not unrelated now, see travis log:
https://travis-ci.org/JuliaGraphs/LightGraphs.jl/jobs/521110145

**Comment by matbesancon** [comment]
@abhinavmehndiratta I think there was really an error in the tests, try to test locally

**Comment by matbesancon** [comment]
sorry nevermind, seems ok

**Comment by abhinavmehndiratta** [comment]
> @abhinavmehndiratta I think there was really an error in the tests, try to test locally

The tests sometimes fail because QR algorithm does not converge for some reason, it isn't related to this I think.

**Pull request merged**